src/plugins/telemetry_server/telemetry_server_impl.cpp

```diff
@@ -0,0 +1,369 @@
+        &msg,
+        home.latitude_deg * 1E7,
+        home.longitude_deg * 1E7,
+        static_cast<int>(home.absolute_altitude_m) * 1E-3,
+        0, // Local X
+        0, // Local Y
+        0, // Local Z
+        q, // surface normal transform
+        NAN, // approach x
+        NAN, // approach y
+        NAN, // approach z
+        get_boot_time() // TO-DO: System boot
+    );
+
+    return _parent->send_message(msg) ? TelemetryServer::Result::Success :
+                                        TelemetryServer::Result::Unsupported;
+}
+
+TelemetryServer::Result TelemetryServerImpl::publish_raw_gps(
+    TelemetryServer::RawGps raw_gps, TelemetryServer::GpsInfo gps_info)
+{
+    mavlink_message_t msg;
+    mavlink_msg_gps_raw_int_pack(
+        _parent->get_own_system_id(),
+        _parent->get_own_component_id(),
+        &msg,
+        raw_gps.timestamp_us,
+        static_cast<uint8_t>(gps_info.fix_type),
+        static_cast<uint32_t>(raw_gps.latitude_deg) * 1E7,
+        static_cast<uint32_t>(raw_gps.longitude_deg) * 1E7,
+        static_cast<uint32_t>(raw_gps.absolute_altitude_m) * 1E3,
+        static_cast<uint32_t>(raw_gps.hdop) * 1E2,
+        static_cast<uint32_t>(raw_gps.vdop) * 1E2,
+        static_cast<uint32_t>(raw_gps.velocity_m_s) * 1E2,
+        static_cast<uint32_t>(raw_gps.cog_deg) * 1E2,
+        gps_info.num_satellites,
+        static_cast<uint32_t>(raw_gps.altitude_ellipsoid_m) * 1E3,
+        static_cast<uint32_t>(raw_gps.horizontal_uncertainty_m) * 1E3,
+        static_cast<uint32_t>(raw_gps.vertical_uncertainty_m) * 1E3,
+        static_cast<uint32_t>(raw_gps.velocity_uncertainty_m_s) * 1E3,
+        static_cast<uint32_t>(raw_gps.heading_uncertainty_deg) * 1E5,
+        static_cast<uint32_t>(raw_gps.yaw_deg) * 1E2);
+
+    return _parent->send_message(msg) ? TelemetryServer::Result::Success :
+                                        TelemetryServer::Result::Unsupported;
+}
+
+TelemetryServer::Result TelemetryServerImpl::publish_battery(TelemetryServer::Battery battery)
+{
+    mavlink_message_t msg;
+
+    uint16_t voltages[10] = {0};
+    uint16_t voltages_ext[4] = {0};
+    voltages[0] = static_cast<uint16_t>(battery.voltage_v) * 1E3;
+
+    mavlink_msg_battery_status_pack(
+        _parent->get_own_system_id(),
+        _parent->get_own_component_id(),
+        &msg,
+        0,
+        MAV_BATTERY_FUNCTION_ALL,
+        MAV_BATTERY_TYPE_LIPO,
+        INT16_MAX,
+        voltages,
+        -1, // TODO publish all battery data
+        -1,
+        -1,
+        static_cast<uint16_t>(battery.remaining_percent) * 1E2,
+        0,
+        MAV_BATTERY_CHARGE_STATE_UNDEFINED,
+        voltages_ext,
+        MAV_BATTERY_MODE_UNKNOWN,
+        0);
+
+    return _parent->send_message(msg) ? TelemetryServer::Result::Success :
+                                        TelemetryServer::Result::Unsupported;
+}
+
+TelemetryServer::Result
+TelemetryServerImpl::publish_status_text(TelemetryServer::StatusText status_text)
```

Comment:
You might want to re-use this: https://github.com/mavlink/MAVSDK/blob/main/src/core/mavlink_statustext_handler.h
But maybe for now it's ok.